packages/main/src/components/Modals/index.tsx

```diff
@@ -0,0 +1,186 @@
+import React, { createRef, ElementType, RefObject } from 'react';
+import ReactDOM from 'react-dom';
+import {
+  Dialog,
+  DialogDomRef,
+  DialogPropTypes,
+  Popover,
+  PopoverDomRef,
+  PopoverPropTypes,
+  ResponsivePopover,
+  ResponsivePopoverDomRef,
+  ResponsivePopoverPropTypes,
+  Toast,
+  ToastDomRef,
+  ToastPropTypes
+} from '../../webComponents';
+import { MessageBox, MessageBoxPropTypes } from '../MessageBox';
+import { ThemeProvider } from '../ThemeProvider';
+
+type ModalReturnType<DomRef> = {
+  ref: RefObject<DomRef>;
+};
+
+type ClosableModalReturnType<DomRef> = ModalReturnType<DomRef> & {
+  close: () => void;
+};
+
+const getContainer = (givenContainer: HTMLElement) => {
+  if (givenContainer) {
+    return givenContainer;
+  }
+  const container = document.createElement('div');
+  container.setAttribute('data-ui5wcr-modal', '');
+  document.body.appendChild(container);
+  return container;
+};
+
+const unmountComponent = (container: HTMLElement) => {
+  ReactDOM.unmountComponentAtNode(container);
+  if (container.hasAttribute('data-ui5wcr-modal')) {
+    container.remove();
+  }
+};
+
+const popupOnAfterCloseFactory = (props, container) => (event) => {
+  if (typeof props.onAfterClose === 'function') {
+    props.onAfterClose(event);
+  }
+  unmountComponent(container);
+};
+
+const render = (Element: ElementType, props: any, container: HTMLElement): Promise<void> => {
+  return new Promise((resolve) => {
```

Comment:
Can I ask why is needed to create a `Promise` and render the element inside a `setTimeout`? 
Also, I wanted to ask you if could be an idea to use a [Portal](https://reactjs.org/docs/portals.html) in this circumstance.

Reply:
Good catch with the Promise - this was some earlier experiment which is removed in the meantime. So I've removed the promise implementation as well.
I'm using `setTimeout` here because ReactDOM.render seems to be blocking other react events when called synchronously, so I decided to make this asynchronous.

We can't use `createPortal` here because it doesn't really render the element into the DOM but is rather an escape hatch from the current DOM structure while rendering.

Reply:
Makes sense, thank for the clarifications.

Reply:
You're welcome 🙂 
And I have to correct myself: the promise is needed here:
https://github.com/SAP/ui5-webcomponents-react/blob/bdc42404876827dfd7bdaea1510bf68e42460501/packages/main/src/components/Modals/index.tsx#L175-L180